dali/operators/numba_function/numba_func.cc

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include "dali/operators/numba_function/numba_func.h"
+
+namespace dali {
+
+DALI_SCHEMA(experimental__NumbaFunc)
+  .DocStr(R"code(Invokes a compiled Numba function passed as a pointer.
+
+.. note::
+    This operator is experimental and its API might change without notice.
+    
+)code")
+  .NumInput(1)
+  .NumOutput(1)
+  .Unserializable()
+  .AddArg("fn_ptr", R"code(Numba function pointer.
+  
+The function should be a Numba C callback function (annotated with cfunc)
+
+  types.void(
+    types.CPointer(OUTPUT_DTYPE), # Pointer to output sample
+    types.CPointer(types.int64),  # Pointer to output sample shape
+    types.CPointer(INPUT_DTYPE),  # Pointer to input sample
+    types.CPointer(types.int64),  # Pointer to input sample shape
+    types.int32                   # Number of sample dimensions
+  )
+
+)code", DALI_INT64)
+  .AddOptionalArg<int>("setup_fn", R"code(Pointer to function which should return output shape.
+
+If the setup function isn't provided then it's assumed output shape and type is the same as input.
+  
+The function should be a Numba C callback function (annotated with cfunc) with the following function signature::
+
+  types.void(
+    types.CPointer(types.int64), # Output batch shape pointer
+    types.int32,                 # Number of dimensions in output shape
+    types.CPointer(types.int32), # Output dtype pointer
+    types.CPointer(types.int64), # Input batch shape pointer
+    types.int32,                 # Number of dimensions in input shape
+    types.int32,                 # Input dtype
+    types.int32,                 # Number of samples in the batch
+    types.int32,                 # Number of outputs
+    types.int32                  # Number of inputs
+  )
+
+)code", 0);
+
+template <typename Backend>
+NumbaFunc<Backend>::NumbaFunc(const OpSpec &spec) : Base(spec) {
+  fn_ptr_ = spec.GetArgument<uint64_t>("fn_ptr");
+  setup_fn_ = spec.GetArgument<uint64_t>("setup_fn");
```

Comment:
I wonder if we can validate these function pointers anyhow.

Reply:
We can only check if fn_ptr_ isn't null.